### PR TITLE
Chart Installation Client and Extensions

### PR DESCRIPTION
**Pull request by igomez06**
**Framework Implementations**
Generated Catalog client for monitoring when an app/chart is deployed.
Implemented Rest calls to install and uninstall a chart
Implemented Extension for rancher monitoring chart, with its clean up.

### CODE REVIEW

File: tests/framework/clients/rancher/catalog/clusterrepo.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	rancherChartsURL = "v1/catalog.cattle.io.clusterrepos/rancher-charts"
+	rancherAppsURL   = "v1/catalog.cattle.io.apps/"
+)
+
+// GetLastestChartVersion is used to get the lastest version of `chartName`
+func (c *Client) GetLastestChartVersion(chartName string) (string, error) {
```
Comment (paynejacob) [inline tests/framework/clients/rancher/catalog/clusterrepo.go:19]:
```suggestion
func (c *Client) GetLatestChartVersion(chartName string) (string, error) {
```

File: tests/framework/clients/rancher/catalog/clusterrepo.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	if len(specifiedChartEntries) < 1 {
+		return "", fmt.Errorf("failed to find chart %s from the chart repo", chartName)
+	}
+
+	lastestVersion := specifiedChartEntries[0].(map[string]interface{})["version"].(string)
+	return lastestVersion, nil
+}
+
+// ChartInstall installs the chart according to the paratmeter `chart`
+func (c *Client) ChartInstall(chart *types.ChartInstallAction) error {
```
Comment (paynejacob) [inline tests/framework/clients/rancher/catalog/clusterrepo.go:45]:
```suggestion
func (c *Client) InstallChart(chart *types.ChartInstallAction) error {
```

File: tests/framework/clients/rancher/catalog/clusterrepo.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+		AbsPath(rancherChartsURL).Param("action", "install").
+		VersionedParams(&metav1.CreateOptions{}, scheme.ParameterCodec).
+		Body(bodyContent).
+		Do(context.TODO())
+
+	return result.Error()
+}
+
+// ChartUninstall uninstalls the chart according to `chartNamespace` and `chartName`
+func (c *Client) ChartUninstall(chartName, chartNamespace string) error {
```
Comment (paynejacob) [inline tests/framework/clients/rancher/catalog/clusterrepo.go:61]:
```suggestion
func (c *Client) UninstallChart(chartName, chartNamespace string) error {
```

File: tests/framework/clients/rancher/catalog/clusterrepo.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	uninstallAction := &types.ChartUninstallAction{
```
Comment (paynejacob) [inline tests/framework/clients/rancher/catalog/clusterrepo.go:62]: why not expose this object? these could influence the tests.  You could split it into 2 methods if you want to keep a simple option available. `SimpleUninstallChart` and `UninstallChart`

File: tests/framework/extensions/charts/ranchermonitoring.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+		Version:     version,
+		Values: v3.MapStringInterface{
+			"global": map[string]interface{}{
+				"cattle": map[string]string{
+					"clusterId":             clusterName,
+					"clusterName":           clusterName,
+					"rkePathPrefix":         "",
+					"rkeWindowsPathPrefix":  "",
+					"systemDefaultRegistry": "",
+					"url":                   "https://israelhatest.qa.rancher.space",
```
Comment (paynejacob) [inline tests/framework/extensions/charts/ranchermonitoring.go:45]: this should not be hard coded.

File: tests/framework/extensions/charts/ranchermonitoring.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+				},
+			},
+			"global": map[string]interface{}{
+				"cattle": map[string]string{
+					"clusterId":             clusterName,
+					"clusterName":           clusterName,
+					"rkePathPrefix":         "",
+					"rkeWindowsPathPrefix":  "",
+					"systemDefaultRegistry": "",
+					"url":                   "https://israelhatest.qa.rancher.space",
```
Comment (paynejacob) [inline tests/framework/extensions/charts/ranchermonitoring.go:73]: this should not be hard coded.

File: tests/framework/extensions/charts/ranchermonitoring.go [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+const (
+	RancherMonitoringNamespace    = "cattle-monitoring-system"
+	RancherMonitoringName         = "rancher-monitoring"
+	RancherMonitoringCRDName      = "rancher-monitoring-crd"
+	RancherMonitoringConfigSecret = "alertmanager-rancher-monitoring-alertmanager"
+)
+
+//InstallRancherMonitoringChart is a helper function that installs the rancher-montitoring chart.
+func InstallRancherMonitoringChart(client *rancher.Client, projectID, clusterName, version string) error {
+	chartInstall0 := types.ChartInstall{
```
Comment (paynejacob) [inline tests/framework/extensions/charts/ranchermonitoring.go:29]:
```suggestion
	chartInstallCRD := types.ChartInstall{
```

File: tests/framework/extensions/charts/ranchermonitoring.go [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+
+const (
+	RancherMonitoringNamespace    = "cattle-monitoring-system"
+	RancherMonitoringName         = "rancher-monitoring"
+	RancherMonitoringCRDName      = "rancher-monitoring-crd"
+	RancherMonitoringConfigSecret = "alertmanager-rancher-monitoring-alertmanager"
+)
+
+//InstallRancherMonitoringChart is a helper function that installs the rancher-montitoring chart.
+func InstallRancherMonitoringChart(client *rancher.Client, projectID, clusterName, version string) error {
```
Comment (paynejacob) [inline tests/framework/extensions/charts/ranchermonitoring.go:28]: Out of curiosity is there a particular reason we picked monitoring to be out test chart? it is one of the heavier ones.  Something like rancher backup or cis would be much faster to test with.

**Review by paynejacob — CHANGES_REQUESTED** [review]

File: tests/framework/clients/rancher/catalog/clusterrepo.go (hunk 1 above)
Comment (igomez06) [reply tests/framework/clients/rancher/catalog/clusterrepo.go]: fixed

File: tests/framework/clients/rancher/catalog/clusterrepo.go (hunk 2 above)
Comment (igomez06) [reply tests/framework/clients/rancher/catalog/clusterrepo.go]: fixed

File: tests/framework/clients/rancher/catalog/clusterrepo.go (hunk 3 above)
Comment (igomez06) [reply tests/framework/clients/rancher/catalog/clusterrepo.go]: fixed

File: tests/framework/clients/rancher/catalog/clusterrepo.go (hunk 4 above)
Comment (igomez06) [reply tests/framework/clients/rancher/catalog/clusterrepo.go]: I decided to keep just the one `UninstallChart` with the `ChartUninstallAction` exposed.

File: tests/framework/extensions/charts/ranchermonitoring.go (hunk 5 above)
Comment (igomez06) [reply tests/framework/extensions/charts/ranchermonitoring.go]: Fixed

File: tests/framework/extensions/charts/ranchermonitoring.go (hunk 6 above)
Comment (igomez06) [reply tests/framework/extensions/charts/ranchermonitoring.go]: Fixed

File: tests/framework/extensions/charts/ranchermonitoring.go (hunk 7 above)
Comment (igomez06) [reply tests/framework/extensions/charts/ranchermonitoring.go]: Fixed

File: tests/framework/extensions/charts/ranchermonitoring.go (hunk 8 above)
Comment (igomez06) [reply tests/framework/extensions/charts/ranchermonitoring.go:28]: It was one of the charts that Brandon suggest to me to try, also it has been implemented in one of our python validation tests. No other reason besides that.

**Review by paynejacob — APPROVED** [review]

**Pull request merged**